libmamba/src/specs/version.cpp

```diff
@@ -217,78 +216,108 @@ namespace mamba::specs
          * ``[1, 2, 0, 0]`` are considered equal, however ``[1, 2]`` and ``[1, 0, 2]`` are not.
          * Similarily ``[1, 1] is less than ``[1, 2, 0]`` but more than ``[1, 1, -1]``
          * because ``-1 < 0``.
+         *
+         * @return The comparison between the two sequence
+         * @return The first index where the two sequence diverge.
          */
-        template <typename Iter1, typename Iter2, typename T, typename Cmp>
+        template <typename Iter1, typename Iter2, typename Empty1, typename Empty2, typename Cmp>
         constexpr auto lexicographical_compare_three_way_trailing(
-            Iter1 first1,
-            Iter1 last1,
-            Iter2 first2,
-            Iter2 last2,
-            T empty,
+            const Iter1 first1,
+            const Iter1 last1,
+            const Iter2 first2,
+            const Iter2 last2,
+            const Empty1 empty1,
+            const Empty2 empty2,
             Cmp comp
-        ) -> strong_ordering
+        ) -> std::pair<strong_ordering, std::size_t>
         {
-            for (; (first1 != last1) && (first2 != last2); ++first1, ++first2)
+            assert(first1 <= last1);
+            assert(first2 <= last2);
```

Comment:
This restricts the implementation to random access iterators only

Reply:
How does the std lib algo deal with this? Should we simply not check anything? I know this could be part of the contract, but I think it can be nice to have extra checks (seeing how tricky it is right now to debug in libmamba).

How about `assert(std::distance(first1, last1) >=0 );` ?

Reply:
I don't think the STL implementation checks anything; I agree that extra checks are a nice-to-have, the `assert(std::distance(first1, last1) >=0 );` akes it work for any iterator so let's go for it.